middleware/csrf_test.go

```diff
@@ -13,14 +12,193 @@ import (
+			whenTokenLookup: "form:csrf",
+			givenCSRFCookie: "token",
+			givenMethod:     http.MethodPost,
+			givenFormTokens: map[string][]string{
+				"csrf": {"invalid_token"},
+			},
+			expectError: "code=403, message=invalid csrf token",
+		},
+		{
+			name:            "nok, missing token from POST form",
+			whenTokenLookup: "form:csrf",
+			givenCSRFCookie: "token",
+			givenMethod:     http.MethodPost,
+			givenFormTokens: map[string][]string{},
+			expectError:     "code=400, message=missing csrf token in the form parameter",
+		},
+		{
+			name:            "ok, token from POST header",
+			whenTokenLookup: "", // will use defaults
+			givenCSRFCookie: "token",
+			givenMethod:     http.MethodPost,
+			givenHeaderTokens: map[string][]string{
+				echo.HeaderXCSRFToken: {"token"},
+			},
+		},
+		{
+			name:            "ok, token from POST header, second token passes",
+			whenTokenLookup: "header:" + echo.HeaderXCSRFToken,
+			givenCSRFCookie: "token",
+			givenMethod:     http.MethodPost,
+			givenHeaderTokens: map[string][]string{
+				echo.HeaderXCSRFToken: {"invalid", "token"},
+			},
+		},
+		{
+			name:            "nok, invalid token from POST header",
+			whenTokenLookup: "header:" + echo.HeaderXCSRFToken,
+			givenCSRFCookie: "token",
+			givenMethod:     http.MethodPost,
+			givenHeaderTokens: map[string][]string{
+				echo.HeaderXCSRFToken: {"invalid_token"},
+			},
+			expectError: "code=403, message=invalid csrf token",
+		},
+		{
+			name:              "nok, missing token from POST header",
+			whenTokenLookup:   "header:" + echo.HeaderXCSRFToken,
+			givenCSRFCookie:   "token",
+			givenMethod:       http.MethodPost,
+			givenHeaderTokens: map[string][]string{},
+			expectError:       "code=400, message=missing csrf token in request header",
+		},
+		{
+			name:            "ok, token from PUT query param",
+			whenTokenLookup: "query:csrf-param",
+			givenCSRFCookie: "token",
+			givenMethod:     http.MethodPut,
+			givenQueryTokens: map[string][]string{
+				"csrf-param": {"token"},
+			},
+		},
+		{
+			name:            "ok, token from PUT query form, second token passes",
+			whenTokenLookup: "query:csrf",
+			givenCSRFCookie: "token",
+			givenMethod:     http.MethodPut,
+			givenQueryTokens: map[string][]string{
+				"csrf": {"invalid", "token"},
+			},
+		},
+		{
+			name:            "nok, invalid token from PUT query form",
+			whenTokenLookup: "query:csrf",
+			givenCSRFCookie: "token",
+			givenMethod:     http.MethodPut,
+			givenQueryTokens: map[string][]string{
+				"csrf": {"invalid_token"},
+			},
+			expectError: "code=403, message=invalid csrf token",
+		},
+		{
+			name:             "nok, missing token from PUT query form",
+			whenTokenLookup:  "query:csrf",
+			givenCSRFCookie:  "token",
+			givenMethod:      http.MethodPut,
+			givenQueryTokens: map[string][]string{},
+			expectError:      "code=400, message=missing csrf token in the query string",
+		},
```

Comment:
I did not find a testcase testing multiple token lookup, maybe I misunderstood rhe intention here.
But a testcase verifying the syntax would be good to have to ensure future compatibility.